Add chatzilena et al as a replication example

SamuelBrand1 opened the pull request:
This PR replicates two parts of the demonstration of epidemic inference using stan from [_Chatzilena et al 2019_](https://www.sciencedirect.com/science/article/pii/S1755436519300325):

- The "deterministic" SIR model fit to a classic boarding school influenza data set. This is a SIR model with Poisson errors.
- The "Stochastic" SIR model fit to same data. This is a SIR model with an autoregressive process on the log-residuals.

In both cases, I use a combination of SciML and EpiAware functionality which worked well.

Closes #446 

## Aside comments

The prior predictive checking (not done in Chatzilena et al) reveals very poor prior specification in the sense that I don't think anyone would reasonably expect the prior marginal median for numbers of ill children. 

This means that I use a combination of MLE and MAP optimisation to start the NUTS sampling in reasonable parts of parameter space; in practise I would change the prior specifications but I felt this deviated from replication/comparison.

Doing the pre-sampling optimisation means that we are getting to reasonable ess is about 10-60x faster than reported in the paper (acknowledging here that the paper is 4 years old and other confounding factors). However, without this optimization the method is prone to stochastic failures. Therefore, pre-sampling steps are quite important to consider going forwards if we use this kind of approach in applications.

seabbs commented:
> Therefore, pre-sampling steps are quite important to consider going forwards if we use this kind of approach in applications.

I think this is a generally true observation in most real-world use cases. Do we already have a initialisation method that samples from the prior (I forget but can check)

seabbs commented:
I see the following Documenter error: 


Code:
  begin
      using EpiAware.EpiAwareUtils: censored_pmf, censored_cdf, ∫F
      using Random, Distributions, StatsBase #utilities for random events
      using DataFramesMeta #Data wrangling
      using CairoMakie, PairPlots #plotting
      using Turing #PPL
  end

Error:
  ArgumentError: Package StatsBase not found in current path.
  - Run `import Pkg; Pkg.add("StatsBase")` to install the StatsBase package.
  Stacktrace:
  
  FYI I think we nearly critically need to move towards pluto wrapped notebooks vs the doc level environment as this kind of thing is easy to do.

seabbs commented:
I think you need to add this to pages.jl (perhaps we can automate this at a later date to avoid this step)

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+# Single population SIR model
+
+As mentioned in _Chatzilena et al_ disease spread is frequently modelled in terms
+of ODE-based models. The study population is divided into compartments representing a specific stage of the epidemic status. In this case, susceptible, infected, and recovered individuals.
```
is this line break intentionnal?

seabbs replied:
ha fair. One option we could take going forward is to put all new sentences on a new line. Makes it easier to see changes in git and maybe plays nicer with precommit. Downside is very boring.

SamuelBrand1 replied:
For a person with a taut readable style this would be no problem...

SamuelBrand1 replied:
Ok I've done this except for bullets

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+> `(du, u, p, t) -> nothing`
+
+Where:
+- `du` is the _vector field_ of the ODE problem, e.g. ${dS \over dt}$, ${dI \over dt}$ etc. This is calculated _in-place_.
```
```suggestion
- `du` is the _vector field_ of the ODE problem, e.g. ${dS \over dt}$, ${dI \over dt}$ etc. This is calculated _in-place_ (commonly denoted using ! in function names in Julia).
```

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+
+# ╔═╡ f16eb00b-2d77-45df-b767-757fe2f5674c
+md"
+We combine vector field function `sir!` with a initial condition `u0` and the integration period `tspan` to make an `ODEProblem`. We do not define the parameters, these will be defined within an inference approach.
```
Suggest add a link out to `EpiProblem` concept we use.

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+
+There was a brief, but intense, outbreak of Influenza within the (semi-) closed community of a boarding school reported to the British medical journal in 1978. The outbreak lasted from 22nd January to 4th February and it is reported that one infected child started the epidemic and then it spread rapidly. Of the 763 children at the boarding scholl, 512 became ill.
+
+We downloaded the data of this outbreak using the R package `outbreaks` which is maintained as part of the [R Epidemics Consortium(RECON)](http://www. repidemicsconsortium.org).
```
maintained is doing a lot of work here

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+N = 763;
+
+# ╔═╡ bb07a580-6d86-48b3-a79f-d2ed9306e87c
+sir_prob = ODEProblem(
```
is it useful here to talk a bit about the ODEProblem object or link somewhere that does

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+
+# ╔═╡ 81501c84-5e1f-4829-a26d-52fe00503958
+md"
+Now we can write the observation model using the `Turing` PPL.
```
is it not the probabliistic or full model (which contains the obs model)

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+        verbose = false)
+
+    ##log-like accumulation using obs##
+    λt = N * sol[2, :] .+ upjitter #expected It
```
is the jitter needed here I thought we had handled that with the obserror types?

SamuelBrand1 replied:
No not the negative safety (unless I missed something v possbile)

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+
+    ##log-like accumulation using obs##
+    λt = N * sol[2, :] .+ upjitter #expected It
+    @submodel obsYt = generate_observations(obs, Yt, λt)
```
obsYt and following doesn't really follow our variable naming conventions. I really don't think this is a big deal FYI but if want to change to _ I would do so

SamuelBrand1 replied:
Good point, I think its worth being consistent here.

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+md"
+We instantiate the model in two ways:
+
+1. `deterministic_mdl`: This conditions the generative model on the data observation. We can sample from this model to find the posterior distribution of the parameters.
```
Again suggest link out to other docs we have

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+We do three things to mitigate this:
+
+1. We choose a switching ODE solver which switches between explicit (`Tsit5`) and implicit (`Rosenbrock23`) solvers. This helps avoid the ODE solver failing when the sampler tries extreme parameter values. This is the default `solver = AutoTsit5(Rosenbrock23())` above.
+2. To avoid the effect of numerically negative small values of `λt` we add a small `upjitter`.
```
sounds like my above question answer is no

SamuelBrand1 replied:
I've got a bit mixed up... is this about not using `apply_method` I'm going to make a comment about this at the bottom.

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+
+1. We choose a switching ODE solver which switches between explicit (`Tsit5`) and implicit (`Rosenbrock23`) solvers. This helps avoid the ODE solver failing when the sampler tries extreme parameter values. This is the default `solver = AutoTsit5(Rosenbrock23())` above.
+2. To avoid the effect of numerically negative small values of `λt` we add a small `upjitter`.
+3. We locate the maximum likelihood point, that is we ignore the influence of the priors, as a useful starting point for `NUTS`.
```
another good place for a callout to elsewhere in our docs

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+"
+
+# ╔═╡ 8d96db67-de3b-4704-9f54-f4ed50a4ecff
+nmle_tries = 100
```
this makes me sad. Can't we do this using our EpiProblem framework? I think this might be out of scope for this PR and be a new issue as I can see it needing a feature?

seabbs replied:
I think the answer is to write the infection model as a EpiAware infection model and then do directly?

seabbs replied:
This would currently be a bit painful so maybe needs to wait for a refactor of the infection submodels

SamuelBrand1 replied:
That is a really good point:

1) https://github.com/CDCgov/Rt-without-renewal/issues/306 We have MAP/MLE opt as an issue
2) I could do this with the pathfinder functionality we already have

seabbs replied:
As this works as is lets leave and update in an issue. I like the idea that we support multiple options. I think it makes sense though to also link out to the idea we have lots of options when we update this.

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+"
+
+# ╔═╡ 71a26408-1c26-46cf-bc72-c6ba528dfadd
+ar = AR(
```
could be cool here to put in a prior predictive check of this sub model

SamuelBrand1 replied:
yes indeed!

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+)
+
+    ##Priors##
+    β ~ LogNormal(0.0, 1.0)
```
wow what if this was rewritten as an infection EpiAware infection model (new issue/PR)

SamuelBrand1 replied:
I would love that.

seabbs replied:
Shall we make a new issue for this? I don't think this is to be addressed here? Then we can close this and linked issues.

seabbs commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+    ##Sample the log-residual AR process##
+    nobs = length(Yt)
+    @submodel κₜ = generate_latent(logobsprob, nobs)
+    λt = @. N * sol[2, :] * exp(κₜ) + upjitter
```
Can we be pro and use any of the helpers (transformLatent eetc etc) if no can we have an issue to make one that would let us do this?

SamuelBrand1 replied:
Yes, really good shout

SamuelBrand1 commented:
Hey, @damonbayer . Was just wondering if you have any thoughts on whether this is a good intro to SciML + EpiAware inference?

SamuelBrand1 commented on EpiAware/docs/src/showcase/replications/chatzilena-2019/index.jl:
```diff
@@ -0,0 +1,622 @@
+# ╔═╡ 9309f7f8-0896-4686-8bfc-b9f82d91bc0f
+@model function stochastic_ode_mdl(Yt, ts, logobsprob, obs, prob, N;
+        solver = AutoTsit5(Rosenbrock23()),
+        upjitter = 1e-2#0.1,
```
Good catch

SamuelBrand1 commented:
I think I've addressed the comments from @seabbs . 

An issue to note, which will be an issue to address, is that in an experiment with using the `ManyPathfinder` -> `NUTS` helper functions (not committed) I found that it made inference both slower (approx x4 but I didn't fully benchmark) and one of the chains for the stochastic model diverged to a nonsense zone of parameter space. This is out of scope for this PR but needs followed up.

SamuelBrand1 commented:
Pertinent to above and @seabbs comment on using `upjitter`. Using the softplus transform instead of an upjitter (or clamp to do relu) stabilises the sampling. Also, using `LogExpFunctions.xexpy` to implement the effect of the AR process on observations in the stochastic model.

After these changes, an experiment to not use initial values to start NUTS still converges onto the correct posterior distributions, so at least in this context this is a big improvement in the stability of the model. I know @damonbayer has wrangled with this in Bayesian inference of ODEs so I wonder what your thoughts on using special functions to stabilise sampling are? 

I've left the initial condition finding steps in the note because it still improves things a bit, and I think is a good example.

seabbs reviewed (approved):
This looks great now. Love it. Looking forward to v2 once we had added more integration to EpiAware.
